swap/protocol.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// ErrEmptyAddressInSignature is used when the empty address is used for the chequebook in the handshake
+var ErrEmptyAddressInSignature = errors.New("empty address in handshake")
+
+// ErrInvalidHandshakeMsg is used when the message received during handshake does not conform to the
+// structure of the HandshakeMsg
+var ErrInvalidHandshakeMsg = errors.New("invalid handshake message")
+
+// Spec is the swap protocol specification
+var Spec = &protocols.Spec{
+	Name:       "swap",
+	Version:    1,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		HandshakeMsg{},
+		EmitChequeMsg{},
+	},
+}
+
+// Protocols is a node.Service interface method
+func (s *Swap) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     s.run,
+		},
+	}
+}
+
+// APIs is a node.Service interface method
+func (s *Swap) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "swap",
+			Version:   "1.0",
+			Service:   NewAPI(s),
+			Public:    false,
+		},
+	}
+}
+
+// Start is a node.Service interface method
+func (s *Swap) Start(server *p2p.Server) error {
+	log.Info("Swap service started")
+	return nil
+}
+
+// Stop is a node.Service interface method
+func (s *Swap) Stop() error {
+	return nil
+}
+
+// verifyHandshake verifies the chequebook address transmitted in the swap handshake
+func (s *Swap) verifyHandshake(msg interface{}) error {
+	handshake, ok := msg.(*HandshakeMsg)
+	if !ok {
+		return ErrInvalidHandshakeMsg
+	}
+
+	if (handshake.ContractAddress == common.Address{}) {
+		return ErrEmptyAddressInSignature
+	}
+
+	return s.verifyContract(context.TODO(), handshake.ContractAddress)
+}
+
+// run is the actual swap protocol run method
+func (s *Swap) run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	protoPeer := protocols.NewPeer(p, rw, Spec)
+
+	handshake, err := protoPeer.Handshake(context.TODO(), &HandshakeMsg{
+		ContractAddress: s.owner.Contract,
+	}, s.verifyHandshake)
+	if err != nil {
+		return err
+	}
+
+	response, ok := handshake.(*HandshakeMsg)
+	if !ok {
+		return ErrInvalidHandshakeMsg
+	}
+
+	beneficiary, err := s.getContractOwner(context.TODO(), response.ContractAddress)
+	if err != nil {
+		return err
+	}
+
+	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, response.ContractAddress)
+	s.addPeer(swapPeer)
+	defer s.removePeer(swapPeer)
+
+	return swapPeer.Run(swapPeer.handleMsg)
+}
+
+func (s *Swap) removePeer(p *Peer) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.peers, p.ID())
+}
+
+func (s *Swap) addPeer(p *Peer) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.peers[p.ID()] = p
+}
+
+func (s *Swap) getPeer(id enode.ID) (*Peer, error) {
+	var err error
+	peer := s.peers[id]
+	if peer == nil {
+		err = fmt.Errorf("peer %s not found", id.String())
+	}
+	return peer, err
+}
+
+type swapAPI interface {
```

Comment:
why do you need an unexported interface for?

Reply:
see #1650.